lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,537 @@
+      if (newChild.isAtEnd())
+        ++portIt;
+      else
+        break;
+    }
+    child = portIt == portEnd ? ChildIterator() : newChild;
+  }
+
+public:
+  explicit InstanceNodeIterator(InstanceOp instance, Node node,
+                                bool end = false)
+      : NodeIterator(node, true), instance(instance) {
+    assert(instance == getValue().getDefiningOp<InstanceOp>() &&
+           "instance must be the defining op of the node value");
+    if (end)
+      return;
+
+    // Query the combinational paths between IOs of the current instance.
+    auto module = getInstanceGraph()->getReferencedModule(instance);
+    auto &combPaths = getCombPathsMap()->FindAndConstruct(module).second;
+    auto &ports = combPaths[getValue().cast<OpResult>().getResultNumber()];
+
+    portEnd = ports.end();
+    portIt = ports.begin();
+    skipToNextValidPort();
+  }
+
+  InstanceNodeIterator &operator++() {
+    if (!child.isAtEnd())
+      ++child;
+    if (child.isAtEnd()) {
+      ++portIt;
+      skipToNextValidPort();
+    }
+    return *this;
+  }
+
+private:
+  InstanceOp instance;
+  SmallVectorImpl<size_t>::iterator portEnd;
+  SmallVectorImpl<size_t>::iterator portIt;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// SubfieldNodeIterator class
+//===----------------------------------------------------------------------===//
+
+namespace {
+class SubfieldNodeIterator : public NodeIterator {
+public:
+  explicit SubfieldNodeIterator(SubfieldOp subfield, Node node,
+                                bool end = false)
+      : NodeIterator(node, true) {
+    assert(subfield == node.value.getDefiningOp<SubfieldOp>() &&
+           "subfield must be the defining op of the node value");
+    if (end)
+      return;
+
+    auto memory = subfield.input().getDefiningOp<MemOp>();
+    if (!memory) {
+      subfield->emitOpError("input must be a port of a MemOp, please run "
+                            "-firrtl-lower-types first");
+      return;
+    }
+
+    auto portKind =
+        memory.getPortKind(subfield.input().cast<OpResult>().getResultNumber());
+    // Combinational path exists only when the current subfield is `addr` and
+    // the read latency is zero.
+    if (subfield.fieldIndex() != 0 || memory.readLatency() != 0 ||
+        portKind == MemOp::PortKind::Write)
+      return;
+
+    // Only `data` or `rdata` subfield is combinationally connected to `addr`
+    // subfield. Find the corresponding subfield op.
+    auto users = subfield.input().getUsers();
+    for (auto it = users.begin(), e = users.end(); it != e; ++it) {
+      auto currentSubfield = dyn_cast<SubfieldOp>(*it);
+      if (!currentSubfield) {
+        it->emitOpError("MemOp must be used by SubfieldOp, please run "
+                        "-firrtl-lower-types first");
+        return;
+      }
+
+      if ((currentSubfield.fieldIndex() == 3 &&
```

Comment:
Is the field index "3" stable?  If so, can we use an enum defined in FIRRTLOps.h instead of a magic number?

Reply:
Yes, they are fixed, according to [this discussion](https://github.com/llvm/circt/pull/1462#discussion_r674460293). Your idea sounds great. Making it a enum could avoid many unnecessary string or magic number comparison.

Reply:
I'll open another PR for this.